src/dataflow-types/src/reconciliation/command.rs

```diff
@@ -0,0 +1,361 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+//! Functionality to reconcile commands between ADAPTER and COMPUTE.
+
+use crate::client::{
+    Client, Command, ComputeCommand, ComputeInstanceId, ComputeResponse, Response, StorageCommand,
+    StorageResponse, TimestampBindingFeedback,
+};
+use anyhow::Error;
+use async_trait::async_trait;
+use mz_expr::GlobalId;
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, VecDeque};
+use timely::progress::frontier::MutableAntichain;
+use timely::progress::ChangeBatch;
+use tracing::warn;
+
+/// Reconcile commands targeted at a COMPUTE instance.
```

Comment:
It would help me to have a more clear understanding of what this means, operationally. For example, I could imagine: "the command and response streams are such that as soon as a `CreateInstance` command is `send`ed, all subsequent `recv` responses will reflect updates as if to a restarted instance; reconnecting clients should ignore all `recv` responses before they send the `Create` command for the corresponding entity, and perhaps shouldn't call `recv` at all until they have reinitialized."

I'm making that up and maybe it is wrong, but .. e.g. stuff like that makes it a bit clearer what is meant to happen. Totally fine that it doesn't exist yet, but I think we want it for the PR to know if it is not-even-wrong or not.

Reply:
Added some words at the top of the file. Will take another look.

Reply:
It's a bit difficult to get this right due to the asynchronous nature of the command/response protocol. I think the client is not in a good position to interrupt the protocol correctly when no coordinator is connected. Currently, it doesn't even know that no coordinator is connected, so it cannot even observe this situation and make decisions based on it.

Assuming it did know, we still have to solve the problem of what happens to replies that we sent while the coordinator was already disconnected. Currently, they're queued and release to the next coordinator once it connects.
* Can we safely drop these messages, assuming their state changes have been applied locally?
* Or should the coordinator be resilient against unsolicited responses and safely ignore them?

I currently believe the second option is simpler to implement.

Also, the client currently has no way to determine that a response has been successfully delivered, due to the properties of TCP. I'm not sure the protocol has been vetted enough to prove that this is fine :)